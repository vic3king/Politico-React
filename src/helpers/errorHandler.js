const errorHandler = obj => {
  let error;

  if (typeof obj === 'string') {
    error = obj;
  }

  if (typeof obj === 'object') {
    Object.values(obj).forEach(err => {
      error = err;
    });
  }

  return error;
};

export default errorHandler;
