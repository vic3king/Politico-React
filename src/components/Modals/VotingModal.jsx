import React, { Component } from 'react';
import Notifications, { notify } from 'react-notify-toast';
import Loader from '../shared/Loader/Loader';
import Votes from '../../services/votes';
import errorHandler from '../../helpers/errorHandler';
import capitalizer from '../../helpers/capitalizer';
import '../../style/modal.scss';

class VotingModal extends Component {
  constructor(props) {
    super(props);
    this.state = {
      candidates: [],
      loading: false,
    };
  }

  hide = e => {
    e.preventDefault();
    const { hide } = this.props;
    hide();
  };

  componentDidMount = async () => {
    this.setState({ loading: true });
    const { officeId } = this.props;
    const candidates = await Votes.getCandidatesByOffice(officeId);

    if (candidates.data.length < 1) {
      this.setState({ loading: false });
      notify.show('No candidates available', 'error');
    }

    this.setState({ candidates: candidates.data });

    if (candidates.status >= 400) {
      this.setState({ loading: false });
      notify.show(errorHandler(candidates.error), 'error');
    }

    if (candidates.status === 200) {
      this.setState({ loading: false });
    }
  };

  onButtonSubmit = async event => {
    event.preventDefault();
    this.setState({ loading: true });
    const { officeId, hide } = this.props;

    const office = Number(officeId);
    const candidate = Number(event.target.id);
    const formDetails = {
      office,
      candidate,
    };

    const vote = await Votes.voteCandidate(formDetails);

    if (vote.status >= 400) {
      this.setState({ loading: false });
      notify.show(errorHandler(vote.error.message), 'error');
    }

    if (vote.status === 201) {
      this.setState({ loading: false });
      hide();
      notify.show('success');
    }
  };

  render() {
    const { loading, candidates } = this.state;

    const candidatesList = candidates.map(candidate => (
      <div className="result-details" key={candidate.id}>
        <div className="profile__image__candidates" />
        <h6>
          Candidate: {capitalizer(candidate.firstname)}{' '}
          {capitalizer(candidate.lastname)}
        </h6>
        <h6>Party: {capitalizer(candidate.partyname)}</h6>
        <h6>Position: {candidate.officename}</h6>
        <button
          id={candidate.id}
          type="submit"
          className="vote-btn"
          onClick={this.onButtonSubmit}
        >
          Vote
        </button>
      </div>
    ));
    return (
      <React.Fragment>
        <Notifications />
        {loading && <Loader />}
        <div className="modal__results">
          <div className="modal-results">
            {candidatesList}
            <button type="button" id="btn" onClick={this.hide}>
              X
            </button>
          </div>
        </div>
      </React.Fragment>
    );
  }
}

export default VotingModal;
