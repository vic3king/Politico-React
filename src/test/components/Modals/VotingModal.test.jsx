import React from 'react';
import { shallow } from 'enzyme';
import { notify } from 'react-notify-toast';
import VotingModal from '../../../components/Modals/VotingModal';
import Votes from '../../../services/votes';
import errorHandler from '../../../helpers/errorHandler';

jest.mock('../../../services/votes');
jest.mock('react-notify-toast');

const hide = jest.fn();
const officeId = 1;

let wrapper;
describe('VotingModal component', () => {
  beforeEach(() => {
    wrapper = shallow(<VotingModal hide={hide} officeId={officeId} />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('hide method', () => {
    wrapper = shallow(<VotingModal hide={hide} officeId={officeId} />);
    const instance = wrapper.instance();
    let event;
    it('should call preventDefault on event', async () => {
      event = {
        preventDefault: jest.fn(),
      };

      await instance.hide(event);
      expect(event.preventDefault).toBeCalled();
    });
  });

  describe('componentDidMount method', () => {
    let instance;
    describe('Api call success', () => {
      beforeAll(() => {
        Votes.getCandidatesByOffice = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 200,
            data: [
              {
                office: '',
              },
            ],
          })
        );
      });

      beforeEach(() => {
        instance = wrapper.instance();
      });

      it('should set loading state to false after successfull api call is made', async () => {
        await instance.componentDidMount();

        expect(instance.state.loading).toBeFalsy();
      });

      it('should toast error when successful but no result exists', async () => {
        Votes.getCandidatesByOffice = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 200,
            data: [],
          })
        );
        await instance.componentDidMount();
        expect(notify.show).toBeCalledWith('No candidates available', 'error');
      });
    });
  });

  describe('Api call 400 errors', () => {
    let instance;
    beforeAll(() => {
      Votes.getCandidatesByOffice = jest.fn().mockImplementation(() =>
        Promise.resolve({
          status: 400,
          error: 'error',
        })
      );
    });

    beforeEach(() => {
      instance = wrapper.instance();
    });

    it('should set loading state to false on validation error', async () => {
      await instance.componentDidMount();
      expect(instance.state.loading).toBeFalsy();
    });

    it('should toast the error message', async () => {
      await instance.componentDidMount();
      expect(notify.show).toBeCalledWith(errorHandler('error'), 'error');
    });
  });

  //   describe('Api call 400 errors', () => {
  //     beforeAll(() => {
  //       Votes.getCandidatesByOffice = jest.fn().mockImplementation(() =>
  //         Promise.resolve({
  //           status: 400,
  //           error: {
  //             message: 'validation error',
  //           },
  //         })
  //       );
  //     });
  //     beforeEach(() => {
  //       instance = wrapper.instance();
  //       event = {
  //         preventDefault: jest.fn(),
  //       };
  //     });

  //     it('should set loading state to false on validation error', async () => {
  //       await instance.onButtonSubmit(event);
  //       expect(instance.state.loading).toBeFalsy();
  //     });

  //     it('should toast the error message', async () => {
  //       await instance.onButtonSubmit(event);
  //       expect(notify.show).toBeCalledWith(
  //         errorHandler('validation error'),
  //         'error'
  //       );
  //     });
  //   });
  // });

  describe('onButtonSubmit method', () => {
    let instance;
    let event;

    describe('Api call success', () => {
      beforeAll(() => {
        Votes.voteCandidate = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 201,
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
        event = {
          preventDefault: jest.fn(),
          target: { id: 2 },
        };
      });

      it('should call preventDefault on event', async () => {
        await instance.onButtonSubmit(event);

        expect(event.preventDefault).toBeCalled();
      });
      it('should call toast the right message on api call success', async () => {
        await instance.onButtonSubmit(event);
        expect(notify.show).toBeCalledWith('success');
      });
      it('should set loading state to false after successfull api call is made', async () => {
        await instance.onButtonSubmit(event);
        expect(instance.state.loading).toBeFalsy();
      });
      it('should call the hide method to close the modal after success', async () => {
        await instance.onButtonSubmit(event);
        expect(hide).toBeCalled();
      });
    });

    describe('Api call 400 errors', () => {
      beforeAll(() => {
        Votes.voteCandidate = jest.fn().mockImplementation(() =>
          Promise.resolve({
            status: 400,
            error: {
              message: 'validation error',
            },
          })
        );
      });
      beforeEach(() => {
        instance = wrapper.instance();
        event = {
          preventDefault: jest.fn(),
          target: { id: 2 },
        };
      });

      it('should set loading state to false on validation error', async () => {
        await instance.onButtonSubmit(event);
        expect(instance.state.loading).toBeFalsy();
      });

      it('should toast the error message', async () => {
        await instance.onButtonSubmit(event);
        expect(notify.show).toBeCalledWith(
          errorHandler('validation error'),
          'error'
        );
      });
    });
  });
});
