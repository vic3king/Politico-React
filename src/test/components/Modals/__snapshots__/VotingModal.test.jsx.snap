// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VotingModal component should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <VotingModal
    hide={[MockFunction]}
    officeId={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <$class
          options={Object {}}
        />,
        <Loader />,
        <div
          className="modal__results"
        >
          <div
            className="modal-results"
          >
            <button
              id="btn"
              onClick={[Function]}
              type="button"
            >
              X
            </button>
          </div>
        </div>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "options": Object {},
        },
        "ref": null,
        "rendered": null,
        "type": [MockFunction],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <div
            className="modal-results"
          >
            <button
              id="btn"
              onClick={[Function]}
              type="button"
            >
              X
            </button>
          </div>,
          "className": "modal__results",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              Array [],
              <button
                id="btn"
                onClick={[Function]}
                type="button"
              >
                X
              </button>,
            ],
            "className": "modal-results",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "X",
                "id": "btn",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": "X",
              "type": "button",
            },
          ],
          "type": "div",
        },
        "type": "div",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <$class
            options={Object {}}
          />,
          <Loader />,
          <div
            className="modal__results"
          >
            <div
              className="modal-results"
            >
              <button
                id="btn"
                onClick={[Function]}
                type="button"
              >
                X
              </button>
            </div>
          </div>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "options": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": [MockFunction],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <div
              className="modal-results"
            >
              <button
                id="btn"
                onClick={[Function]}
                type="button"
              >
                X
              </button>
            </div>,
            "className": "modal__results",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                Array [],
                <button
                  id="btn"
                  onClick={[Function]}
                  type="button"
                >
                  X
                </button>,
              ],
              "className": "modal-results",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "X",
                  "id": "btn",
                  "onClick": [Function],
                  "type": "button",
                },
                "ref": null,
                "rendered": "X",
                "type": "button",
              },
            ],
            "type": "div",
          },
          "type": "div",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
